broker-core/src/main/java/io/zeebe/broker/workflow/state/ElementInstanceState.java

```diff
@@ -327,6 +327,29 @@ public void updateFailedToken(IndexedRecord indexedRecord) {
     return records;
   }
 
+  @FunctionalInterface
+  public interface TokenVisitor {
+    void visitToken(IndexedRecord indexedRecord);
+  }
+
+  public void visitFailedTokens(long scopeKey, TokenVisitor visitor) {
+    visitTokens(scopeKey, Purpose.FAILED_TOKEN, visitor);
+  }
+
+  private void visitTokens(long scopeKey, Purpose purpose, TokenVisitor visitor) {
```

Comment:
This method is similar to `getTokenEvents(...)`. Can we delete the other one?

Reply:
I refactored a bit the method. I tried also to avoid creating garbage in the getToken method, but this was in the end not easy possible. I would like to postpone this until we refactor the whole stuff.

Reply:
https://github.com/zeebe-io/zeebe/issues/1691 :)